### PR TITLE
Added 'debug' and 'opt' keys to graph.json

### PR DESCRIPTION
**Pull request by adusca**
We now have to main keys: 'opt' and 'debug', each key contains it's own dictionary of platforms and tests. The new output is: https://pastebin.mozilla.org/8825005

Writing this patch I discovered that some test jobs in allthethings.json have no upstream builders. Originally I thought it was a bug in platforms.py, but I inspected they manually in Treeherder and found out they really did not have upstream builders (like b2g_ubuntu64_vm gaia-try opt test gaia-js-integration-7 and Android 4.0 Panda mozilla-esr31 opt test mochitest-6).

### CODE REVIEW

**Comment by armenzg** [comment]
Hi Alice,
The data looks very nice.

Could you please add another parameter to build_tests_per_platform_graph()?
I want to add repo_name="try" since in general we only care about creating this graphs for try.
If anyone ever finds a use where we also care about other repositories we should figure it out at that time.

With regards to some builders not having upstream builders, it is a bug on the Release Engineering side.
They generate their confins in buildbot-configs, however, build builders and test builders are somewhat disconnected logics and depend on the humans getting the configs right.
We could benefit from a script that could help us create a list of **test** builders that don't have an upstream **build** builder. That way we can then go to Release Engineering to please go and disable those builders.

For instance, look at esr31:
https://treeherder.mozilla.org/#/jobs?repo=mozilla-esr31&filter-searchStr=Android&exclusion_profile=false
There are no Android jobs at all.

Gaia-try is special, very special.
https://treeherder.mozilla.org/#/jobs?repo=gaia-try
If you click on the different jobs and look at the left of the bottom sidebard that pops up, you will see that they all have the same "Request time" (both build and test jobs).
I believe the way it works is that all jobs are triggered at the same time.
The build jobs will behave as normal, however, we won't test them.
The test jobs will grab an already existant build and packaged tests:
http://hg.mozilla.org/integration/gaia-try/raw-file/713aa5da49ffb7745ea48757580a72d05b828cac/linux64.json
That json file is updated every time there is a new PR on the Github gaia repo:
http://hg.mozilla.org/integration/gaia-try/log/713aa5da49ff/linux64.json
I think a bot determines what is the latest gaia builds:
https://ftp.mozilla.org/pub/mozilla.org/b2g/tinderbox-builds/mozilla-central-linux64_gecko/?C=M;O=D

The long story short is that all jobs for gaia-try are to be considered upstream builders.

There might be some other exceptions so please let me know for any oddities you find.

**Comment by armenzg** [comment]
There are also some **build** platforms that are not showing up.
Trychooser [1] allows me to select the following:
linux,linux64,linux64-asan,linux64-st-an,linux64-valgrind,linux64-br-haz,linux64-b2g-haz,linux64-mulet,macosx64,macosx64-mulet,win32,win32-mulet,win64,android-api-9,android-api-11,android-x86,emulator,emulator-jb,emulator-kk,emulator-l,linux32_gecko,linux64_gecko,macosx64_gecko,linux64-sh-haz

Could you please give me a list of builders that do not fit in any of our current expectation? I want to help figure out why. There are too many special conditions! :dancer: 

[1] http://trychooser.pub.build.mozilla.org

**Comment by armenzg** [comment]
I forgot to mention that it might be because some of those builds do not trigger any tests.

File: mozci/platforms.py [hunk 1]
```diff
@@ -151,16 +147,32 @@ def _get_test(buildername):
...
-    graph = collections.defaultdict(list)
+    graph = {}
+    graph['debug'] = collections.defaultdict(list)
+    graph['opt'] = collections.defaultdict(list)
 
     for builder in builders:
         if is_downstream(builder):
-            platform = get_associated_platform_name(builder)
+            # Some builders in allthethings are not triggered by any
+            # upstream, so we have to keep this part in a try.
```
Comment (armenzg) [inline mozci/platforms.py]: Do you have a builder that is an example of this?

File: mozci/platforms.py [hunk 2]
```diff
@@ -151,16 +147,32 @@ def _get_test(buildername):
...
+
             test = _get_test(builder)
-            if test not in graph[platform]:
-                graph[platform].append(test)
+            if platform.endswith('-debug'):
+                key = 'debug'
+                platform = platform[:-len('-debug')]
+
+            else:
+                key = 'opt'
```
Comment (armenzg) [inline mozci/platforms.py]: There is a chance that we might need to keep the **whole** key for the sake of other things in the future but for now let's go with it.

**Comment by armenzg** [comment]
Once we switch to only producing the mapping for try, could we append after each key the builder that it represents?
We would probably have to add a "downstream" key for build jobs and a "buildername" key.
This could be beneficial when building a TryChooser UI to be able to show hints to the user as to which builders will be getting added.
It will also be useful for issue #109

**Comment by adusca** [comment]
I'm investigating the missing platforms. We have in our graph:
Debug platforms: 
android-api-11,  emulator,  linux,  linux64,  linux64_gecko,  macosx64,  win32,  win64
Opt platforms:
android-api-11,  android-api-9,  android-x86,  emulator,  linux,  linux-pgo,  linux32_gecko,  linux64,  linux64-asan,  linux64-cc,  linux64-mulet,  linux64-pgo,  linux64_gecko,  macosx64,  win32,  win32-pgo,  win64,  win64-pgo.

The strings "linux64-br-haz" , "linux64-b2g-haz",  "linux64-sh-haz", "emulator-l", "win32-mulet", do not show up anywhere on allthethings.json.

Only cedar builds have 'emulator-kk' and 'emulator-jb' as platforms.

linux-valgrind is a more interesting case. There are no builders with "platform" : "linux64-valgrind", but some builders (like  "Linux x86-64 b2g-inbound valgrind") have the string in their shortname. Turns out the tryparser uses a prettyName dictionary and -valgrind is added to some builders [1].

The only macosx64-multet builder in try is "OS X Mulet try build" and there are no tests triggered by it.

[1] http://mxr.mozilla.org/build/source/buildbotcustom/misc.py#1208

**Comment by armenzg** [comment]
If you want, we can just focus on the repo_name; that way there is less things to consider.

With regards to this,

> The strings "linux64-br-haz" , "linux64-b2g-haz", "linux64-sh-haz", "emulator-l",
> "win32-mulet", do not show up anywhere on allthethings.json.
> We can hardcode it on mozci until we fix the issue in allthethings.json
> 
> Only cedar builds have 'emulator-kk' and 'emulator-jb' as platforms.
> Generally, A-team members enable jobs in there to get them fixed before enabling them in other places.
> In this case, I think these should be disabled. I will take care of it.

Let's only consider valgrind for try.

Considering only the "try" repo, we should have the following builds which I believe don't have any tests triggered:

NOTE: Only Linux Mulet currently have tests.

$ grep "haz" all_builders.txt | grep "try"
 u'b2g_try_linux64-b2g-haz_dep',
 u'linux64-br-haz_try_dep',
 u'linux64-sh-haz_try_dep',
MacAir ~ $ grep "emulator-" all_builders.txt | grep "try"
 u'b2g_try_emulator-debug_dep',
 u'b2g_try_emulator-jb-debug_dep',
 u'b2g_try_emulator-jb_dep',
 u'b2g_try_emulator-kk-debug_dep',
 u'b2g_try_emulator-kk_dep',
 u'b2g_try_emulator-l-debug_dep',
 u'b2g_try_emulator-l_dep',
MacAir ~ $ grep "Mulet" all_builders.txt | grep "try"
 u'Linux x86-64 Mulet try build',
 u'OS X Mulet try build',
 u'Ubuntu VM 12.04 x64 Mulet try opt test mochitest-1',
 u'Ubuntu VM 12.04 x64 Mulet try opt test mochitest-2',
 u'Ubuntu VM 12.04 x64 Mulet try opt test mochitest-3',
 u'Ubuntu VM 12.04 x64 Mulet try opt test mochitest-4',
 u'Ubuntu VM 12.04 x64 Mulet try opt test mochitest-5',
 u'Ubuntu VM 12.04 x64 Mulet try opt test reftest-1',
 u'Ubuntu VM 12.04 x64 Mulet try opt test reftest-2',
 u'Ubuntu VM 12.04 x64 Mulet try opt test reftest-3',
 u'Ubuntu VM 12.04 x64 Mulet try opt test reftest-4',
 u'Ubuntu VM 12.04 x64 Mulet try opt test reftest-5',
 u'Ubuntu VM 12.04 x64 Mulet try opt test reftest-6',
 u'Win32 Mulet try build',
MacAir ~ $ grep "valgrind" all_builders.txt | grep "try"
 u'Linux x86-64 try valgrind',

**Comment by adusca** [comment]
Since build_tests_per_platform_graph() already receives a list of builders as input I wrote a function filter_builders_matching() and I pass its output (with keyword = " try ") to build_tests_per_platform_graph().

The current output looks like: https://pastebin.mozilla.org/8825164 . Every platform is a dictionary with one key for every build job and a key 'tests'. Under it build job there is a list of test jobs triggered by it and under 'tests' is a list of every test that runs in that platform.

Should build jobs that don't trigger anything and platforms without any tests also be in the graph?

**Comment by armenzg** [comment]
This is looking great!

> Should build jobs that don't trigger anything and platforms without any tests also be in the graph?
> Yes, please. Could you please point me to which ones you're referring to? (just in case :smiley: )

File: mozci/platforms.py [hunk 3]
```diff
@@ -149,18 +145,45 @@ def _get_test(buildername):
     return buildername.split(" ")[-1]
 
 
+def filter_builders_matching(builders, keyword):
```
Comment (armenzg) [inline mozci/platforms.py]: Let's make this a private function for now.

File: mozci/platforms.py [hunk 4]
```diff
@@ -149,18 +145,45 @@ def _get_test(buildername):
...
+    """Find all the builders in a list that contain a keyword."""
+    return [builder for builder in builders if keyword in builder]
```
Comment (armenzg) [inline mozci/platforms.py]: I think you can use lambda in here.

**Comment by adusca** [comment]
The build jobs in try that do not trigger any test jobs are:

Android armv7 API 9 try debug build
Linux try leak test spidermonkey_try-arm-sim build
Linux try leak test spidermonkey_try-warnaserrdebug build
Linux try spidermonkey_try-warnaserr build
Linux x86-64 try debug asan build
Linux x86-64 try debug static analysis build
Linux x86-64 try leak test spidermonkey_try-exactrooting build
Linux x86-64 try leak test spidermonkey_try-generational build
Linux x86-64 try leak test spidermonkey_try-rootanalysis build
Linux x86-64 try leak test spidermonkey_try-warnaserrdebug build
Linux x86-64 try spidermonkey_try-warnaserr build
Linux x86-64 try valgrind
OS X 10.7 64-bit try debug static analysis build
OS X Mulet try build
WINNT 5.2 try leak test spidermonkey_try-generational build
WINNT 5.2 try leak test spidermonkey_try-warnaserrdebug build
WINNT 5.2 try spidermonkey_try-generational build
WINNT 5.2 try spidermonkey_try-warnaserr build
WINNT 6.1 x86-64 try leak test spidermonkey_try-generational build
WINNT 6.1 x86-64 try leak test spidermonkey_try-warnaserrdebug build
WINNT 6.1 x86-64 try spidermonkey_try-generational build
WINNT 6.1 x86-64 try spidermonkey_try-warnaserr build
WINNT 6.1 x86-64 try xulrunner nightly
Win32 Mulet try build

**Comment by armenzg** [comment]
Hi Alice,
I commited this but is not currently working:
scripts/misc/generate_list_of_testers_without_builders.py

It probably needs the changes in this PR request.

Once you addres my previous comments, could you please check that this works?

Do the builders you mention in https://github.com/armenzg/mozilla_ci_tools/pull/112#issuecomment-78316871, show up when generating graphs.json?

I think some of those builders should be deleted.

**Comment by adusca** [comment]
Every one of these builders shows up in the current graph.json with an empty list of triggered tests: https://pastebin.mozilla.org/8825312

Do you mean delete them from the graph or from allthethings.json?

**Comment by armenzg** [comment]
> Do you mean delete them from the graph or from allthethings.json?
> I will be deleting this from buildbot-configs (which will be updated in allthethings.json)
>  WINNT 6.1 x86-64 try xulrunner nightly

graphs.json looks good as-is. Thanks!

**Comment by armenzg** [comment]
I landed this for now:
https://github.com/armenzg/mozilla_ci_tools/commit/2a8289a156538a1ea18aa53971578ab9c99213f6

I also filed issue #124 with regards to L10n jobs. We should skip them until we fix this.

**Comment by armenzg** [comment]
Let me know when we're ready to review the PR and merge it.

I meant this:
https://github.com/armenzg/mozilla_ci_tools/commit/5e2de316479990b9caf7187c87c7dd5c337e1847

**Comment by adusca** [comment]
I think I'm done.

**Comment by armenzg** [comment]
Thank you so much!
